libraries/tools/kotlin-symbol-processing-api/src/org/jetbrains/kotlin/ksp/utils.kt

```diff
@@ -99,33 +100,69 @@ fun KSDeclaration.isOpen() = !this.isLocal()
-        && !this.modifiers.contains(Modifier.PROTECTED)
-        && !this.modifiers.contains(Modifier.INTERNAL)
+fun KSDeclaration.isPublic() = this.getVisibility() == Visibility.PUBLIC
+
+fun KSDeclaration.isProtected() = this.getVisibility() == Visibility.PROTECTED
 
 fun KSDeclaration.isInternal() = this.modifiers.contains(Modifier.INTERNAL)
 
 fun KSDeclaration.isPrivate() = this.modifiers.contains(Modifier.PRIVATE)
 
+fun KSDeclaration.isJavaPackagePrivate() = this.getVisibility() == Visibility.JAVA_PACKAGE
+
+fun KSDeclaration.closestClassDeclaration(): KSClassDeclaration? {
+    if (this is KSClassDeclaration) {
+        return this
+    } else {
+        return this.parentDeclaration?.closestClassDeclaration()
+    }
+}
+
 // TODO: cross module visibility is not handled
+// TODO: generics not handled, need API change to take type argument.
 fun KSDeclaration.isVisibleFrom(other: KSDeclaration): Boolean {
+    fun KSDeclaration.isSamePackage(other: KSDeclaration): Boolean = this.containingFile?.packageName == other.containingFile?.packageName
+
+    // lexical scope for local declaration.
+    fun KSDeclaration.parentDeclarationsForLocal(): List<KSDeclaration> {
+        val parents = mutableListOf<KSDeclaration>()
+
+        var parentDeclaration = this.parentDeclaration!!
+
+        while (parentDeclaration.isLocal()) {
+            parents.add(parentDeclaration)
+            parentDeclaration = parentDeclaration.parentDeclaration!!
+        }
+
+        parents.add(parentDeclaration)
+
+        return parents
+    }
+
+    fun KSDeclaration.isVisibleInPrivate(other: KSDeclaration) =
+        this.parentDeclaration == other.parentDeclaration
+                || this.parentDeclaration == other
+                || (
+                this.parentDeclaration == null
+                        && other.parentDeclaration == null
+                        && this.containingFile == other.containingFile
+                )
+
     return when {
         // locals are limited to lexical scope
-        this.isLocal() -> this.parentDeclaration == other
+        this.isLocal() -> this.parentDeclarationsForLocal().contains(other)
         // file visibility or member
-        this.isPrivate() -> {
-            this.parentDeclaration == other.parentDeclaration
-                    || this.parentDeclaration == other
-                    || (
-                    this.parentDeclaration == null
-                            && other.parentDeclaration == null
-                            && this.containingFile == other.containingFile
-                    )
-        }
+        this.isPrivate() -> this.isVisibleInPrivate(other)
         this.isPublic() -> true
         this.isInternal() && other.containingFile != null && this.containingFile != null -> true
+        this.isJavaPackagePrivate() -> this.isSamePackage(other)
+        // Need to address use case when parent declaration is generic.
```

Comment:
Can you give an example?

Reply:
This is for protected methods, generic subtyping can't be determined without seeing the type arguments, e.g. List<Int> and List<Number> vs List<out Int> and List<out Number>